---
title: "Denoising Method Analysis"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
# Load packages
packages <- c('tidyverse', 'vegan')
lapply(packages, library, character.only = T)
source("code/parse_alpha_maxee.R")
```


## Create dissimilarity matricies for mantel tests

For each ee filtering and alpha value denoising, calculate a dissimilarity matrix of sampling sites based on haplotype abundances.
~~Using the Cao index to account for potentially high beta diversity, and the unknown relationships between sampling and read counts for aquatic sampling of eDNA.~~ USING BINOMIAL METHOD AND CONVERTED TO PRESENCE/ABSCENCE

```{r dissimilarity matricies}
for (folder in dir(pattern = "^denoised_haplotypes")) {
  #read in haplotype data, dropping metadata columns and the duplicated sites 
  #(2 sites end up duplicated somewhere in the denoising pipeline of JAMP)
  haplo_table <- paste0(folder, "/E_haplo_table.csv") %>% 
    read_csv(
      col_names = TRUE, 
      show_col_types = FALSE, 
      col_select = -c("sort", "haplotype", "OTU", "sequences", "CR_216_2_S46_PE.1", "DIA_1_S79_PE.1")) 
  
  #remove last row (site totals)
  haplo_table <- haplo_table[1:nrow(haplo_table) - 1,] 
  
  #transpose the distance matrix so haplotypes are columns and sites are rows.
  haplo_table <- haplo_table %>% t()
  
  # convert to presence absence
  haplo_table[haplo_table > 1] <- 1
  
  #dissimilarity matrix using binomial method (input must be presence/absence).
  vegdist(haplo_table, method = "binomial", binary = FALSE) %>% 
    as.matrix() %>% as.data.frame() %>% 
      #write to file named dist_matrix.csv in appropriate folder.
      write_csv(file = paste0(folder, "/dist_matrix.csv"),col_names = FALSE) 
}
```


## Mantel tests

Compare each dissimilarity matrix to the default settings of ee = 0.5 and unoise_alpha = 5

```{r mantel tests}
default_dist_matrix <- read_csv("./denoised_haplotypes_maxee0_5_alpha_5_2023-08-25/dist_matrix.csv", 
                                col_names = FALSE, show_col_types = FALSE) 

for (folder in dir(pattern = "^denoised_haplotypes")) {
  dist_matrix <- paste0(folder, "/dist_matrix.csv") %>% 
    read_csv(col_names = FALSE, show_col_types = FALSE) 
  
  cat(paste0("mantel test for: ", folder, "\n"), 
      file = "mantel_test_results.txt", append = TRUE)
  
  mantel_result <- mantel(as.dist(default_dist_matrix), as.dist(dist_matrix), 
                          method = "pearson", permutations = 999)
  
  #save results for each comparison to file
  cat("standardized mantel statistic (r-value): ", mantel_result$statistic, 
      "\np-value: ", mantel_result$signif, "\n\n", 
      file = "output_files/mantel_test_results.txt", append = TRUE) 
}
```


## Compare number of haplotypes within OTU's between different denoising methods

Compare how the number of observed OTU's change with each method.
Use OTU centroid sequences to compare the number of ESV's associated with each sequence according to different denoising methods. 

```{r OTU summary table}
haplo2OTU <- list.files(
  pattern = ".*_haplo2OTU.csv", 
  recursive = TRUE, 
  full.names = TRUE) %>% 
    grep("_data", ., value = TRUE)

OTU_seqs <- data.frame(method = character(0), 
                       sequences = character(0), 
                       ESV_count = double(0))

for (file in haplo2OTU) {
  
  file_data <- read_csv(file, 
                        col_select = c("ESV_count", "sequences"), 
                        show_col_types = FALSE)
  
  file_data$method <- substring(
    text = file, 
    first = nchar("../denoised_haplotypes_"), 
    last = nchar(file) - nchar("_2023-08-25/_data/4_denoised/E_haplo2OTU.csv"))
  
  OTU_seqs <- bind_rows(OTU_seqs, file_data)
}

OTU_seqs <- parse_alpha_maxee(OTU_seqs)


```

## OTU Graph

```{r OTU graph}
# Plot number of OTU's for each method
OTU_seqs %>% group_by(alpha, maxee) %>% summarise(count = n()) %>% 
  ggplot((aes(x = alpha, y = count, group = maxee, color = maxee))) +
    geom_line() +
    scale_color_discrete() +
    ggtitle("Number of OTU's based on denoising parameters") +
    ylab("Number of OTU's")
ggsave(filename = "output_files/OTU_graph.png", 
       device = "png")
```

## ESV Graph

``` {r ESV Graph}
# Plot number of ESV's for each method
OTU_seqs %>% group_by(alpha, maxee) %>% summarise(ESV_count = sum(ESV_count)) %>%
  ggplot((aes(x = alpha, y = ESV_count, group = maxee, color = maxee))) +
    geom_line() +
    scale_color_discrete() +
    ggtitle("Number of ESV's based on denoising parameters") +
    ylab("Number of ESV's")
ggsave(filename = "output_files/ESV_graph.png",
       device = "png")
```

# ESV Table

``` {r ESV table}
# Compare number of ESV's per OTU centroid for each method. 
sequence_ESVs_by_method <- OTU_seqs %>% 
  spread(key = "method", value = "ESV_count", fill = 0) %>% 
  select(-c(maxee, alpha))

# Centroid sequences where at least one method had a different number of ESV's
ESV_method_differences <- sequence_ESVs_by_method %>% 
  filter_at(vars(starts_with("maxee")), any_vars(. != maxee0_5_alpha_1))
```
As expected the number of OTU's and the number of ESV's increases as the alpha value is relaxed. The line for ee=0.5 remains lower than the less strict filtering of ee=1.0. 
The number of observed OTU's flattens out quickly as the alpha value increases becoming flat after an alpha of 4. 
The number of ESV's increases until an alpha value of 8. 
This suggests there may be some rare ESV's that are being discarded at lower alpha values, or that these ESV's are actually sequencing errors and were rightfully discarded. 
Regardless, from a species conservation standpoint, it appears that an alpha value of 3 is perfectly acceptable, and relaxing the alpha value will not significantly increase the number of OTU's observed.
When considering species haplotypes looking at phylogeny and population genetics it may be worth considering a more relaxed alpha value of 8. However, this has the risk of including erroneous sequences.

## NMS of the methods in ESV space
```{r NMS, results=FALSE}
# species matrix, columns are haplotypes
ESV_MDS_in <- t(sequence_ESVs_by_method)[-1,]
mode(ESV_MDS_in) <- "numeric"

ESV_MDS_out <- metaMDS(ESV_MDS_in, autotransform = FALSE,
          distance = "bray",
          k = 2,
          maxit = 999, 
          trymax = 500,
          wascores = TRUE)

stressplot(ESV_MDS_out)
plot(ESV_MDS_out, "sites")
orditorp(ESV_MDS_out, "sites")
```

